src/storage/src/source/reclock.rs

```diff
@@ -7,931 +7,651 @@
-/// field.
-///
-/// It effectively turns pairs of source_upper retractions + insertions (placed next to each other
-/// in the stream by integrate into MzOffset diffs
-fn differentiate(
-    updates: &mut Vec<((PartitionId, MzOffset), Timestamp, Diff)>,
-) -> impl Iterator<Item = (PartitionId, Timestamp, MzOffset)> + '_ {
-    // Ensure that the updates are sorted to have updates for the same partition next to each other
-    // and in ascending offset order
-    consolidation::consolidate_updates(updates);
-    let mut prev_update = None;
-    updates
-        .drain(..)
-        .filter_map(move |((pid, offset), ts, diff)| match diff {
-            -1 => {
-                prev_update = Some((pid, offset, ts));
-                None
-            }
-            1 => {
-                let prev_offset = match prev_update.take() {
-                    Some((prev_pid, prev_offset, prev_ts)) => {
-                        assert_eq!(prev_pid, pid, "invalid bindings");
-                        assert_eq!(prev_ts, ts, "invalid bindings");
-                        prev_offset
-                    }
-                    None => MzOffset::from(0),
-                };
-                offset
-                    .checked_sub(prev_offset)
-                    .map(move |diff| (pid, ts, diff))
-            }
-            _ => panic!("invalid binding"),
-        })
+pub struct ReclockFollower<FromTime: Timestamp, IntoTime: Timestamp> {
+    inner: Rc<RefCell<ReclockFollowerInner<FromTime, IntoTime>>>,
 }
 
-/// Integrates the provided updates containing the change of offset in the diff field into updates
-/// containing insertions and retractions of full offset statements.
-fn integrate<'a>(
-    prev_state: &'a HashMap<PartitionId, MzOffset>,
-    updates: &'a [(PartitionId, MzOffset)],
-) -> impl Iterator<Item = ((PartitionId, MzOffset), Diff)> + 'a {
-    updates.into_iter().flat_map(|(pid, diff)| {
-        let (retraction, prev_offset) = match prev_state.get(pid).copied() {
-            Some(prev_offset) => (Some(((pid.clone(), prev_offset), -1)), prev_offset),
-            None => (None, MzOffset::from(0)),
-        };
-
-        let next_offset = prev_offset + *diff;
-        retraction
-            .into_iter()
-            .chain([((pid.clone(), next_offset), 1)])
-    })
-}
-
-struct ReclockFollowerInner {
-    /// A dTVC trace of the remap collection containing all consolidated updates at
-    /// `t` such that `since <= t < upper` indexed by partition and sorted by time.
-    remap_trace: HashMap<PartitionId, Vec<(Timestamp, MzOffset)>>,
+struct ReclockFollowerInner<FromTime: Timestamp, IntoTime: Timestamp> {
+    /// A dTVC trace of the remap collection containing all updates at `t: since <= t < upper`.
+    // NOTE(petrosagg): Once we write this as a timely operator this should just be an arranged
+    // trace of the remap collection
+    remap_trace: Vec<(FromTime, IntoTime, i64)>,
```

Comment:
can we not use `Diff` here?

Reply:
I can make it `Diff`, yeah. It's a bit weird because technically it's not that if `Diff` changed it everything would work fine. But the same time `Diff` won't change. Will change